providers/dns/bunny/bunny.go

```diff
@@ -172,28 +164,52 @@ func (d *DNSProvider) findZone(ctx context.Context, authZone string) (*bunny.DNS
 		return nil, err
 	}
 
+	zone := findZone(zones, authZone)
+	if zone == nil {
+		return nil, fmt.Errorf("could not find DNSZone domain=%s", authZone)
+	}
+
+	return zone, nil
+}
+
+func findZone(zones *bunny.DNSZones, domain string) *bunny.DNSZone {
+	domains := possibleDomains(domain)
+
+	var domainLength int
+
 	var zone *bunny.DNSZone
 	for _, item := range zones.Items {
-		if item != nil && deref(item.Domain) == authZone {
-			zone = item
-			break
+		if item == nil {
+			continue
 		}
-	}
 
-	if zone == nil {
-		return nil, fmt.Errorf("could not find DNSZone zone=%s", authZone)
+		curr := deref(item.Domain)
+
+		if slices.Contains(domains, curr) && domainLength < len(curr) {
+			domainLength = len(curr)
+
+			zone = item
+		}
 	}
 
-	return zone, nil
+	return zone
 }
 
-func getZoneName(fqdn string) (string, error) {
-	authZone, err := dns01.FindZoneByFqdn(fqdn)
-	if err != nil {
-		return "", err
+func possibleDomains(domain string) []string {
+	var domains []string
+
+	labelIndexes := dns.Split(domain)
+
+	for i, index := range labelIndexes {
+		if i == len(labelIndexes)-1 {
```

Comment:
Hm. There are (quite a few) TLDs which don't allow registration domains at the 2nd level (`.uk` was a well-known example). Including `co.uk` in the result slice for `example.co.uk` is still wrong.

I'd use golang.org/x/net/publicsuffix (which leverages https://github.com/publicsuffix/list) here to detect the eTLD.

Reply:
This is just a quick optimization, the goal was not to really skip all possible TLD.